libs/cardano-ledger-binary/src/Cardano/Ledger/TreeDiff.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- | Collect all the basic TreeDiff stuff in one place.
+--   Including orphan instances from external types.
+--   So anyplace this module is imported to get access to the
+--   ToExpr class, one also gets the orphan instances.
+module Cardano.Ledger.TreeDiff
+  ( Expr (App, Rec, Lst),
+    ToExpr (listToExpr, toExpr),
+    defaultExprViaShow,
+    trimExprViaShow,
+    diffExpr,
+    expectExprEqual,
+    expectExprEqualWithMessage,
+    ediffEq,
+  )
+where
+
+import qualified Cardano.Crypto.Hash as Hash
+import Cardano.Crypto.Hash.Class ()
+import Cardano.Slotting.Block (BlockNo)
+import Cardano.Slotting.Slot (EpochNo (..), EpochSize (..), SlotNo (..), WithOrigin (..))
+import Control.Monad (unless)
+import Data.Foldable (toList)
+import Data.IP (IPv4, IPv6)
+import Data.Maybe.Strict (StrictMaybe)
+import Data.Primitive.Types (Prim)
+import Data.Sequence.Strict (StrictSeq)
+import Data.TreeDiff (ansiWlEditExpr)
+import Data.TreeDiff.Class (ToExpr (listToExpr, toExpr), defaultExprViaShow, ediff)
+import Data.TreeDiff.Expr (Expr (App, Lst, Rec))
+import Data.TreeDiff.QuickCheck (ediffEq)
+import Data.VMap (VB, VMap, VP, toMap)
+import Test.Hspec (Expectation, HasCallStack, expectationFailure)
+
+-- =====================================================
+-- Cardano functions that deal with TreeDiff and ToExpr
+
+trimExprViaShow :: Show a => Int -> a -> Expr
+trimExprViaShow n x = App (take n (drop 1 (show x)) ++ "..") []
```

Comment:
What we are trying to do is hide the surrounding " that are introduced by 'show', we use this only for Nonces and Hashes etc. where we really only want to see some short prefix. But I agree since the comparison is done on the whole thing the shortening it has dangers.
It would be nice if there was a constructor of Expr, that was just a string and a number
 Short :: Int -> String -> Expr.
then we could get the best of both worlds on things built with exprViaShow, compare on the string, but only show the first 'n' things when it is displayed. So for now I will just make it a synonym for exprViaShow.